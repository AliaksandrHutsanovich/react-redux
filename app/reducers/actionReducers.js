import {
    addCategory,
    addSubCategory,
    deleteCategory,
    editCategory,
    editTaskStatus,
    addTask,
    editTask,
    deleteTask,
    insertCategory,
    insertTask,
    editCategoryRevive,
    editTaskRevive
} from '../actions/actions';
import { handleActions } from 'redux-actions';
import immutable from 'immutable';
const Map = immutable.Map,
    List = immutable.List;
import { initialState } from './states/initialState';

export const actionReducers = handleActions({

    [addCategory]: {
        next(state, { payload }) {
            return state
                .update("categories",
                (categories) => categories.push(Map({ title: payload.title, subCategories: List([]), tasks: List([]) })));
        }
    },

    [addSubCategory]: {
        next(state, { payload }) {
            return state
                .updateIn(payload.path,
                subCategories => subCategories.push(Map({ title: payload.title, subCategories: List([]), tasks: List([]) })));
        }
    },

    [deleteCategory]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state.deleteIn(fullPath);
        }
    },

    [editCategory]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                category => {
                    return category.set("title", payload.title);
                });
        }
    },

    [editTaskStatus]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                task => {
                    return task.set("isFinished", payload.value);
                });
        }
    },

    [addTask]: {
        next(state, { payload }) {
            return state
                .updateIn(payload.path,
                tasks => tasks.push(Map({ title: payload.title, Description: "", isFinished: false })));
        }
    },

    [editTask]: {
        next(state, { payload }) {
            if (payload.newPath) {
                if (payload.newPathParam) {
                    state = state
                        .updateIn(payload.newPath,
                        tasks => tasks.insert(payload.newPathParam, Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
                }
                else {
                    state = state
                        .updateIn(payload.newPath,
                        tasks => tasks.push(Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
                }
                if (payload.oldPath) {
                    let fullOldPath = Object.assign([], payload.oldPath);
                    fullOldPath.push(payload.oldPathParam);
                    return state.deleteIn(fullOldPath);
                }
                else {
                    return state;
                }
            }
            else {
                let fullOldPath = Object.assign([], payload.oldPath);
                fullOldPath.push(payload.oldPathParam);
                return state
                    .updateIn(fullOldPath,
                    task => task.set('title', payload.title).set('Description', payload.description).set('isFinished', payload.isFinished));
            }
        }
    },

    [deleteTask]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state.deleteIn(fullPath);
        }
    },

    [insertCategory]: {
        next(state, { payload }) {
            return state.updateIn(payload.path, categories => categories.insert(payload.pathParam, Map({ title: payload.title, subCategories: payload.subCategories, tasks: payload.tasks })));
        }
    },

    [insertTask]: {
        next(state, { payload }) {
            return state.updateIn(payload.path, tasks => tasks.insert(payload.pathParam, Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
        }
    },

    [editCategoryRevive]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                category => {
                    return category.set("title", payload.titlePrimary);
                });
        }
    },

    [editTaskRevive]: {
        next(state, { payload }) {
            if (payload.newPath) {
                let fullNewPath = Object.assign([], payload.newPath);
                fullNewPath.push(payload.newPathParam);

                return state
                    .updateIn(payload.oldPath,
                    tasks => tasks.insert(payload.oldPathParam, Map({ title: payload.titlePrimary, Description: payload.descriptionPrimary, isFinished: payload.isFinishedValuePrimary })))
                    .deleteIn(fullNewPath);
            }
            else {
                let fullOldPath = Object.assign([], payload.oldPath);
                fullOldPath.push(payload.oldPathParam);
                return state.updateIn(fullOldPath, task => task.set('title', payload.titlePrimary).set('Description', payload.descriptionPrimary).set('isFinished', payload.isFinishedValuePrimary));
            }
        }
    }

}, Map().merge(initialState));