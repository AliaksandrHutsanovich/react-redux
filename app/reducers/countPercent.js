import { countPercentage, incrementInDone, decrementInDone, incrementInAll, decrementInAll } from '../actions/actions';
import { handleActions } from 'redux-actions';
import immutable from 'immutable';
const Map = immutable.Map;

export const countPer—Åent = handleActions({

    [countPercentage]: {
        next(state, { payload }) {
            return state.set('done', payload.done).set('all', payload.all);
        }
    },

    [incrementInDone]: {
        next(state, { payload }) {
            let done = state.get('done') + 1;
            return state.set('done', done);
        }
    },

    [decrementInDone]: {
        next(state, { payload }) {
            let done = state.get('done') - 1;
            return state.set('done', done);
        }
    },

    [incrementInAll]: {
        next(state, { payload }) {
            let all = state.get('all') + 1;
            return state.set('all', all);
        }
    },

    [decrementInAll]: {
        next(state, { payload }) {
            let all = state.get('all') - 1;
            return state.set('all', all);
        }
    }

}, Map({ done: 0, all: 0 }));

