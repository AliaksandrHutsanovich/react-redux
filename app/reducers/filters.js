import { Map } from 'immutable';
import { handleActions } from 'redux-actions';
import {
  showDoneTasks,
  searchTasks,
} from '../actions';

const filters = handleActions({

  [showDoneTasks]: {
    next(state, { payload }) {
      return state.set('showDone', payload);
    },
  },

  [searchTasks]: {
    next(state, { payload }) {
      return state.set('searchKey', payload);
    },
  },

}, Map({ showDone: true, searchKey: '' }));

export default filters;
