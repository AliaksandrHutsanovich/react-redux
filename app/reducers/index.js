export { default } from './reducer';
export { initialState } from './states/initialState';
