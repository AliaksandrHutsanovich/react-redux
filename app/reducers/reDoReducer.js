import { Map, List } from 'immutable';
import { handleActions } from 'redux-actions';
import { addToReDo, deleteFromReDo, clearReDo } from '../actions';

const reDoReducer = handleActions({

  [addToReDo]: {
    next(state, { payload }) {
      return state
        .update('redoOperations', (operations) => operations.push(Map(payload)));
    },
  },

  [deleteFromReDo]: {
    next(state) {
      return state.update('redoOperations', (operations) => operations.delete(state.get('redoOperations').toArray().length - 1));
    },
  },

  [clearReDo]: {
    next(state) {
      return state.set('redoOperations', List([]));
    },
  },

}, Map({ redoOperations: List([]) }));

export default reDoReducer;
