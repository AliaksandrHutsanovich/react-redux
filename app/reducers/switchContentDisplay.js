import { Map } from 'immutable';
import { handleActions } from 'redux-actions';
import { switchContentDisplay } from '../actions';

const contentDisplay = handleActions({
  [switchContentDisplay]: {
    next(state, { payload: { isDisplayed, url } }) {
      if (!url) {
        return state
          .set('isDisplayed', isDisplayed);
      }
      if (isDisplayed === undefined && url) {
        return state.set('url', url);
      }
      return state
        .set('isDisplayed', isDisplayed)
        .set('url', url);
    },
  },
}, Map({ isDisplayed: true, url: '' }));

export default contentDisplay;
