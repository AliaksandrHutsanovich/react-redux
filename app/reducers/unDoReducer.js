import { Map, List } from 'immutable';
import { handleActions } from 'redux-actions';
import { addToUnDo, deleteFromUnDo } from '../actions/actions';

const unDoReducer = handleActions({

  [addToUnDo]: {
    next(state, { payload }) {
      return state.update('undoOperations', (operations) => operations.push(Map(payload)));
    },
  },

  [deleteFromUnDo]: {
    next(state) {
      return state.update('undoOperations', (operations) => operations.delete(state.get('undoOperations').toArray().length - 1));
    },
  },

}, Map({ undoOperations: List([]) }));

export default unDoReducer;
