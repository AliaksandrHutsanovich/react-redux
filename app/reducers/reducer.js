import { combineReducers } from 'redux';
import {actionReducers} from './actionReducers'; 
import {filters} from  './filters';
import { countPerсent } from './countPercent';
import { reDoReducer } from './reDoReducer';
import { unDoReducer } from './unDoReducer';

export default combineReducers({
    actionReducers,
    filters,
    countPerсent,
    reDoReducer,
    unDoReducer
});








