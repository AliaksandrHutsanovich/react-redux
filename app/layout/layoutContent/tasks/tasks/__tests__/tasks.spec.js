import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';
import { Map } from 'immutable';
import { List } from 'antd';
import { Provider } from 'react-redux';
import Tasks from '../tasks';
import { initialState as store } from '../../../../../reducers';

const initialState = Map().merge(store);
const mockStore = configureStore();

describe('Unit test of tasks component', () => {
  const dispatch = jest.fn();
  const dataStore = mockStore({
    actionReducers: initialState,
    contentDisplay: Map().merge({ isDisplayed: true, url: '' }),
  });
  dataStore.dispatch = dispatch;

  const wrapper = shallow(
    <Tasks
      showDone
      searchKey=""
      match={{ url: 'categories-0' }}
      store={dataStore}
    />,
  );

  it('shallow render test of tasks component', () => {
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    expect(wrapper.dive().shallow().find(List)
      .dive()).toMatchSnapshot();
  });

  it('dispatch should be called in useEffect', async () => {
    await act(async () => mount(
      <Provider store={dataStore}>
        <Tasks
          showDone
          searchKey=""
          match={{ url: 'categories-0' }}
        />
      </Provider>,
    ));
    expect(dispatch).toHaveBeenCalled();
  });
});
