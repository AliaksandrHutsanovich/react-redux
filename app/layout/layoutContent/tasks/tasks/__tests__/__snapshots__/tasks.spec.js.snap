// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test of tasks component shallow render test of tasks component 1`] = `
<div>
  <List
    bordered={false}
    className="list"
    dataSource={
      Array [
        Object {
          "Description": "draw",
          "index": 0,
          "isFinished": true,
          "location": Object {},
          "title": "graph",
        },
        Object {
          "Description": "count",
          "index": 1,
          "isFinished": true,
          "location": Object {},
          "title": "counting",
        },
      ]
    }
    itemLayout="horizontal"
    loading={false}
    pagination={false}
    renderItem={[Function]}
    split={false}
  />
</div>
`;

exports[`Unit test of tasks component shallow render test of tasks component 2`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
